src/app/data-model-provider/MetadataList.cpp

```
@@ -25,106 +25,158 @@ namespace app {
+        mBufferIsAllocated = true;
+        VerifyOrReturnError(new_buffer != nullptr, CHIP_ERROR_NO_MEMORY);
+        memcpy(new_buffer, mBuffer, mElementCount * mElementSize);
+        mBuffer = new_buffer;
+    }
+    mCapacity = mElementCount + numElements;
 
-    mAllocated = true;
-    mCapacity  = numElements;
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR GenericMetadataList::AppendRaw(const void * buffer)
+CHIP_ERROR GenericAppendOnlyBuffer::AppendSingleElementRaw(const void * buffer)
 {
-    VerifyOrReturnError(!mIsImmutable, CHIP_ERROR_INCORRECT_STATE);
-    VerifyOrReturnError(mElementCount < mCapacity, CHIP_ERROR_NO_MEMORY);
+    VerifyOrReturnError(mElementCount < mCapacity, CHIP_ERROR_BUFFER_TOO_SMALL);
     memcpy(mBuffer + mElementCount * mElementSize, buffer, mElementSize);
     mElementCount++;
     return CHIP_NO_ERROR;
 }
 
-void GenericMetadataList::Invalidate()
+CHIP_ERROR GenericAppendOnlyBuffer::AppendElementArrayRaw(const void * buffer, size_t numElements)
 {
-    if ((mBuffer != nullptr) && mAllocated)
+    ReturnErrorOnFailure(EnsureAppendCapacity(numElements));
+
+    memcpy(mBuffer + mElementCount * mElementSize, buffer, numElements * mElementSize);
+    mElementCount += numElements;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericAppendOnlyBuffer::ReferenceExistingElementArrayRaw(const void * buffer, size_t numElements)
+{
+    if (mBuffer == nullptr)
     {
-        chip::Platform::MemoryFree(mBuffer);
-        mBuffer    = nullptr;
-        mAllocated = false;
+        // we can NEVER append with 0 capacity, so const cast is safe
+        mBuffer       = const_cast<uint8_t *>(static_cast<const uint8_t *>(buffer));
+        mElementCount = numElements;
+        // The assertions below are because we know the buffer is null/not allocated yet
+        // Not included because these seem to affect flash size surprisingly enough...
```

What does "not included" mean?

---

Deleted ... I had `assert()` before and that increases flash. When I made VerifyOrDie the increase became negligible (on qpg at least), so I left them in ... seems better to check than silently fail. Good catch!